external-crates/move/crates/move-vm-runtime/src/interpreter/eval.rs

```diff
@@ -0,0 +1,1158 @@
+        ast::{Bytecode, Function},
+        Loader, ModuleDefinitionResolver,
+    },
+    native_functions::NativeContext,
+    trace,
+};
+use fail::fail_point;
+use move_binary_format::{errors::*, file_format::JumpTableInner};
+use move_core_types::{
+    account_address::AccountAddress,
+    gas_algebra::{NumArgs, NumBytes},
+    vm_status::StatusCode,
+};
+use move_vm_config::runtime::{VMConfig, VMRuntimeLimitsConfig};
+#[cfg(feature = "gas-profiler")]
+use move_vm_profiler::GasProfiler;
+use move_vm_profiler::{
+    profile_close_frame, profile_close_instr, profile_open_frame, profile_open_instr,
+};
+use move_vm_types::{
+    data_store::DataStore,
+    gas::{GasMeter, SimpleInstruction},
+    loaded_data::runtime_types::Type,
+    values::{
+        IntegerValue, Reference, Struct, StructRef, VMValueCast, Value, Variant, VariantRef,
+        Vector, VectorRef,
+    },
+};
+use smallvec::SmallVec;
+
+use crate::native_extensions::NativeContextExtensions;
+use std::collections::VecDeque;
+
+use super::state::CallFrame;
+
+macro_rules! set_err_info {
+    ($frame:expr, $e:expr) => {{
+        let function = $frame.function();
+        $e.at_code_offset(function.index(), $frame.pc)
+            .finish($frame.location())
+    }};
+}
+
+struct RunContext<'loader, 'native, 'native_lifetimes> {
+    loader: &'loader Loader,
+    link_context: AccountAddress,
+    defn_resolver: ModuleDefinitionResolver<'loader>,
+    extensions: &'native mut NativeContextExtensions<'native_lifetimes>,
+}
+
+impl RunContext<'_, '_, '_> {
+    // TODO: The Run Context should hold this, not go get it from the Loader.
+    fn vm_config(&self) -> &VMConfig {
+        self.loader.vm_config()
+    }
+}
+
+const DEBUG_STEP_PRINT: bool = false;
+
+/// Main loop for the execution of a function.
+///
+/// This runs a newly-made Machine until it is complete. It expects the Machine to have a current
+/// call frame set up, with no operands on the stack and no existing call stack. It runs in a loop,
+/// calling `step`, until the call stack is empty.current frame of the
+/// This function sets up a `Frame` and calls `execute_code_unit` to execute code of the
+/// function represented by the frame. Control comes back to this function on return or
+/// on call. When that happens the frame is changes to a new one (call) or to the one
+/// at the top of the stack (return). If the call stack is empty execution is completed.
+pub fn run(
+    start_state: MachineState,
+    data_store: &impl DataStore,
+    gas_meter: &mut impl GasMeter,
+    extensions: &mut NativeContextExtensions,
+    loader: &Loader,
+) -> VMResult<Vec<Value>> {
+    let link_context = data_store.link_context();
+    let resolver = start_state.current_frame.resolver(link_context, loader);
+
+    let mut run_context = RunContext {
+        loader,
+        link_context,
+        defn_resolver: resolver,
+        extensions,
+    };
+
+    let mut state = start_state;
+
+    let mut buf = String::new();
```

Comment:
I assume the compiler would remove this but why not put it inside the `if` on both if branches?
Yes we would be allocating a string in a loop but not sure it would matter.
Not a big deal, sorry to even mention and feel free to leave it as is, lol. It's just that it makes me check where the variable is used in the function when it is really local to the debug printing.